### DISCUSSION TITLE
Understanding Differences Between Update, Insert, and Delete Operations

### DISCUSSION THREAD

**Discussion by dongbin86** [vincent-herlemont/native_db#264] [General]
### Question:

I’m trying to clarify the use of update versus insert operations with a practical example, as well as the implications for delete operations, especially when secondary indexes are involved.

For an update operation, I see it as `update(old_item, new_item)`, where I need to retrieve the existing item, modify it, and then update it. For an insert operation, I simply use `insert(item)`. My question is, if I have an entity with a known primary key, and I populate the other fields directly before inserting, will this correctly update the existing record with the same primary key? Additionally, how does this affect any secondary indexes?

For the delete operation, I use `delete(item)`. If the struct for `item` has a `default` method, could I simply set the primary key in the default value and pass it to the delete method? Would this correctly delete the record with that primary key, and would it have any issues with secondary indexes?

#### Example:

Suppose I have an entity with the following fields:

1. `id` (Primary Key)
2. `name`
3. `email`
4. `created_at`
5. `updated_at`
6. `status`
7. `category` (Secondary Index)

If I have an entity where I know the `id` (primary key), and I set `name`, `email`, and `status` before calling `insert(entity)`, will this update the existing record with the same primary key? And, does this have any impact on the `category` secondary index?

For delete, if I create a default entity and only set the `id` field, will calling `delete(entity)` remove the correct record? Would this process have any implications for secondary indexes like `category`?

**Comment by vincent-herlemont** [comment]
I will try to clarify these points; to put it simply:

- [`insert(item)`](https://docs.rs/native_db/latest/native_db/transaction/struct.RwTransaction.html#method.insert) inserts an element **that does not exist**; otherwise, an error is returned.
- [`upsert(item)`](https://docs.rs/native_db/latest/native_db/transaction/struct.RwTransaction.html#method.upsert) inserts an element and updates that element based on primary keys. **All secondary keys are updated** and constraints are checked.
- [`update(itemA, itemB)`](https://docs.rs/native_db/latest/native_db/transaction/struct.RwTransaction.html#method.update) updates `itemA` with `itemB`. `itemA` is found based on its primary key and is completely replaced by `itemB`. **All keys, primary and secondary, are replaced** and constraints are checked.
- [`remove(item)`](https://docs.rs/native_db/latest/native_db/transaction/struct.RwTransaction.html#method.remove) removes the `item`, which is found based on its primary keys.

Does this answer your questions?

**Comment by dongbin86** [comment]
Your answer was very clear. When I first used native_db, it didn't have 'upsert', and inserting an existing item didn’t seem to throw an error. I’ve now upgraded to the latest version. Thanks!

**Reply by vincent-herlemont** [reply]
@dongbin86 Great, if you don't have any other questions on this topic. Could you please mark my response as "solutions"?
